File: testdata/generate_qm35_reference_737p28.m
function meta = generate_qm35_reference_737p28(srcPath, outPath, metaPath)
%GENERATE_QM35_REFERENCE_737P28  Native-rate QM35 code-9 SYNC template.
%
%   META = GENERATE_QM35_REFERENCE_737P28() writes
%     testdata/reference_preamble_code9_737p28.cf32
%     testdata/reference_preamble_code9_737p28_metadata.json
%   from testdata/reference_preamble.bin using the same 48/65 polyphase
%   contract as testdata/generate_qm35_reference_737p28.py.
%
%   The 998.4 MS/s source is the one-SYNC waveform produced by
%   UWB_demodulation/+uwbdecoder/buildUwbReference.m (L2-normalized).
%   Energy, group delay and sample-index base are recorded in META so C++
%   and MATLAB detector starts can be compared within one sample or the
%   documented filter-delay tolerance.
%
%   This file is the MATLAB generator of record.  When MATLAB is absent
%   the Python companion produces bit-identical metadata fields (length,
%   energy, group delay) under the same resample_poly / Kaiser-5 FIR.

    if nargin < 1 || isempty(srcPath)
        here = fileparts(mfilename('fullpath'));
        srcPath = fullfile(here, 'reference_preamble.bin');
    end
    here = fileparts(srcPath);
    if nargin < 2 || isempty(outPath)
        outPath = fullfile(here, 'reference_preamble_code9_737p28.cf32');
    end
    if nargin < 3 || isempty(metaPath)
        metaPath = fullfile(here, 'reference_preamble_code9_737p28_metadata.json');
    end

    fid = fopen(srcPath, 'rb');
    if fid < 0
        error('cannot open %s', srcPath);
    end
    raw = fread(fid, inf, 'float32=>single');
    fclose(fid);
    src = complex(raw(1:2:end), raw(2:2:end));

    up = 48;
    down = 65;
    half = 10;
    native = resample(double(src), up, down);
    native = single(native(:));
    ntaps = 2 * half * max(up, down) + 1;
    gd = (ntaps - 1) / 2 / down;
    energy_before = double(native' * native);
    nrm = norm(double(native)) + eps;
    native = single(double(native) / nrm);
    energy = double(native' * native);

    fid = fopen(outPath, 'wb');
    if fid < 0
        error('cannot write %s', outPath);
    end
    interleaved = zeros(2 * numel(native), 1, 'single');
    interleaved(1:2:end) = real(native);
    interleaved(2:2:end) = imag(native);
    fwrite(fid, interleaved, 'float32');
    fclose(fid);

    meta = struct();
    meta.description = ['One QM35 / IEEE 802.15.4a HRP code-9 SYNC symbol ' ...
        'at the X410 native rate 737.28 MS/s.'];
    meta.code_index = 9;
    meta.preamble_repetitions = 64;
    meta.sfd_mode = '4z2';
    meta.input_file = 'reference_preamble.bin';
    meta.output_file = 'reference_preamble_code9_737p28.cf32';
    meta.input_rate_hz = 998.4e6;
    meta.output_rate_hz = 737.28e6;
    meta.interp = up;
    meta.decim = down;
    meta.resample_half_length = half;
    meta.source_length = numel(src);
    meta.template_length = numel(native);
    meta.energy = energy;
    meta.energy_before_normalize = energy_before;
    meta.group_delay_samples = gd;
    meta.group_delay_domain = 'native_737p28';
    meta.sample_index_base = 0;
    meta.dtype = 'complex64';
    meta.filter_delay_tolerance_samples = 1;
    meta.generator = 'testdata/generate_qm35_reference_737p28.m';
    meta.matlab_reference = 'UWB_demodulation/+uwbdecoder/buildUwbReference.m';

    txt = jsonencode(meta);
    fid = fopen(metaPath, 'w');
    fwrite(fid, txt);
    fclose(fid);
end

File: testdata/generate_qm35_reference_737p28.py
#!/usr/bin/env python3
"""Generate the production 737.28 MS/s QM35 code-9 SYNC template.

Contract (docs/phase1/GROK_X410_QM35盲捕获与自动周期锁定开发方案.md §8):

  testdata/reference_preamble.bin          # 998.4 MS/s, one SYNC, L2-normalized
        -- resample_poly 48/65 -->
  testdata/reference_preamble_code9_737p28.cf32
  testdata/reference_preamble_code9_737p28_metadata.json

Must follow UWB_demodulation/+uwbdecoder/buildUwbReference.m:

  * source waveform is one SYNC symbol
  * energy-normalize with ||x||_2 (MATLAB: x / (norm(x)+eps))
  * sample-index base 0 (template[0] aligns with the first chip sample)

Anti-aliasing is scipy.signal.resample_poly (FIR, Kaiser beta=5, half-length
10 * max(up, down)), the same family of polyphase convention used by the
65/48 host resampler. Group delay is recorded so C++ / MATLAB comparisons
can apply a documented 1-sample-or-filter-delay tolerance.

This script is the generator of record when MATLAB is not available.
The companion .m file implements the same contract.
"""

from __future__ import annotations

import argparse
import json
import os
import sys

import numpy as np
from scipy.signal import resample_poly

HERE = os.path.dirname(os.path.abspath(__file__))
DEFAULT_SRC = os.path.join(HERE, "reference_preamble.bin")
DEFAULT_OUT = os.path.join(HERE, "reference_preamble_code9_737p28.cf32")
DEFAULT_META = os.path.join(HERE, "reference_preamble_code9_737p28_metadata.json")

INPUT_RATE_HZ = 998.4e6
OUTPUT_RATE_HZ = 737.28e6
INTERP = 48
DECIM = 65
# scipy.signal.resample_poly default: numtaps ~= 2 * n + 1 with n = 10 * max(up, down)
RESAMPLE_HALF_LENGTH = 10


def l2_normalize(x: np.ndarray) -> np.ndarray:
    n = float(np.linalg.norm(x))
    if n <= 0.0:
        return x.astype(np.complex64, copy=False)
    return (x / n).astype(np.complex64)


def resample_group_delay_samples(up: int, down: int, half: int) -> float:
    """Output-rate group delay of resample_poly's default FIR.

    The prototype FIR has length 2*half*max(up,down)+1 and is applied at the
    upsampled rate. After decimation the delay in output samples is
    ((ntaps-1)/2) / down.
    """
    ntaps = 2 * half * max(up, down) + 1
    return float(ntaps - 1) / 2.0 / float(down)


def generate(src_path: str, out_path: str, meta_path: str) -> dict:
    src = np.fromfile(src_path, dtype=np.complex64)
    if src.size == 0:
        raise RuntimeError(f"empty source template: {src_path}")
    src_energy = float(np.sum(np.abs(src) ** 2))
    # Source is already L2-normalized by make_reference_template.py / MATLAB.
    native = resample_poly(src, INTERP, DECIM, window=("kaiser", 5.0))
    native = np.asarray(native, dtype=np.complex64)
    gd = resample_group_delay_samples(INTERP, DECIM, RESAMPLE_HALF_LENGTH)
    native_unnorm_energy = float(np.sum(np.abs(native) ** 2))
    native = l2_normalize(native)
    native_energy = float(np.sum(np.abs(native) ** 2))
    native.tofile(out_path)
    meta = {
        "description": (
            "One QM35 / IEEE 802.15.4a HRP code-9 SYNC symbol at the X410 "
            "native rate 737.28 MS/s. Derived from the 998.4 MS/s "
            "buildUwbReference.m template by polyphase 48/65 resample."
        ),
        "code_index": 9,
        "preamble_repetitions": 64,
        "sfd_mode": "4z2",
        "input_file": os.path.relpath(src_path, HERE),
        "output_file": os.path.relpath(out_path, HERE),
        "input_rate_hz": INPUT_RATE_HZ,
        "output_rate_hz": OUTPUT_RATE_HZ,
        "interp": INTERP,
        "decim": DECIM,
        "resample_window": ["kaiser", 5.0],
        "resample_half_length": RESAMPLE_HALF_LENGTH,
        "source_length": int(src.size),
        "source_energy": src_energy,
        "template_length": int(native.size),
        "energy": native_energy,
        "energy_before_normalize": native_unnorm_energy,
        "group_delay_samples": gd,
        "group_delay_domain": "native_737p28",
        "sample_index_base": 0,
        "dtype": "complex64",
        "filter_delay_tolerance_samples": 1,
        "generator": "testdata/generate_qm35_reference_737p28.py",
        "matlab_reference": "UWB_demodulation/+uwbdecoder/buildUwbReference.m",
    }
    with open(meta_path, "w", encoding="utf-8") as f:
        json.dump(meta, f, indent=2, sort_keys=True)
        f.write("\n")
    return meta


def main() -> int:
    p = argparse.ArgumentParser()
    p.add_argument("--src", default=DEFAULT_SRC)
    p.add_argument("--out", default=DEFAULT_OUT)
    p.add_argument("--meta", default=DEFAULT_META)
    args = p.parse_args()
    meta = generate(args.src, args.out, args.meta)
    print(
        "wrote {out}: length={n} energy={e:.9f} group_delay={gd:.6f}".format(
            out=args.out,
            n=meta["template_length"],
            e=meta["energy"],
            gd=meta["group_delay_samples"],
        )
    )
    print("wrote", args.meta)
    return 0


if __name__ == "__main__":
    sys.exit(main())
